### PR TITLE
FIX-#3171: compute shape hint at the QueryCompiler constructor

### LINKED ISSUE #3171: `QueryCompiler.columnarize` triggers unnecessary frame execution
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): f9c99c3817d35f7bf0d91df9aa9644545ea27517
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
from modin.experimental.engines.omnisci_on_ray.test.test_dataframe import (
    set_execution_mode,
)

df = pd.DataFrame({"a": [1, 1, 1, 2, 3], "b": [1, 2, 3, 4, 5]})
set_execution_mode(df, "lazy")

res = df.groupby("a").sum()
res.squeeze(axis=1)  # RuntimeError: unexpected execution triggered on lazy frame
```

<details><summary>Traceback</summary>

```
Traceback (most recent call last):
  File "repr.py", line 14, in <module>
    res.squeeze(axis=1)
  File "/localdisk/dchigare/repos/modin_bp/modin/pandas/dataframe.py", line 1938, in squeeze
    return Series(query_compiler=self._query_compiler)
  File "/localdisk/dchigare/repos/modin_bp/modin/pandas/series.py", line 160, in __init__
    self._query_compiler = query_compiler.columnarize()
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 46, in method_wrapper
    return method(self, *args, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 624, in columnarize
    len(self.index) == 1 and self.index[0] == "__reduced__"
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/backends/omnisci/query_compiler.py", line 336, in _get_index
    return self._modin_frame.index
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/engines/omnisci_on_ray/frame/data.py", line 1170, in _get_index
    self._execute()
  File "/localdisk/dchigare/repos/modin_bp/modin/experimental/engines/omnisci_on_ray/frame/data.py", line 1024, in _execute
    raise RuntimeError("unexpected execution triggered on lazy frame")
RuntimeError: unexpected execution triggered on lazy frame
```

</details>

### Describe the problem
`columnarize` transposes single-row query compilers in order to translate it to the proper Series representation (single-column qc). To determine the shape of the qc without triggering the actual computation, `columnarize` [uses `_shape_hint`](https://github.com/modin-project/modin/blob/f9c99c3817d35f7bf0d91df9aa9644545ea27517/modin/experimental/backends/omnisci/query_compiler.py#L611) field of the self, it appears that sometimes the `shape_hint` is missing (like in the single-column groupby case) and so unnecessary shape computations are triggered.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3171 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py
Comment (dchigarev) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1773]: index at the OmniSci backend is represented by hidden frame's columns, so we need to test shape detection with different kinds of indices to check that it's considering only data cols

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (vnlitvinov) [inline modin/experimental/backends/omnisci/query_compiler.py:89]:
```suggestion
        if shape_hint is None and len(self._modin_frame.columns) == 1:
```
this should work ever so slightly faster in the case when `shape_hint` is already known :)

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1791]: why aren't you passing `index` in this line?

**Review by vnlitvinov — COMMENTED** [review]
Mostly LGTM, want to hear from @modin-project/modin-omnisci before stamping full approval though.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py]: good point, changed as you suggested

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py]: fixed

**Review by devin-petersohn — COMMENTED** [review]
@dchigarev Do you think shape hints would also be useful in modin core?

**Comment by dchigarev** [comment]
@devin-petersohn 
> @dchigarev Do you think shape hints would also be useful in modin core?

As far as I understand, shape-hint is used at OmniSci backend only to indicate whether 1D query compiler is a single-row or a single-column (both of these can perpesent Series object), this information helps to skip shape computation at the [`columnarize`](https://github.com/modin-project/modin/blob/e74cde994bef969abd307ad146e6a5d00847ce51/modin/experimental/backends/omnisci/query_compiler.py#L601) method which [is used](https://github.com/modin-project/modin/blob/e74cde994bef969abd307ad146e6a5d00847ce51/modin/pandas/series.py#L160) at Series constructor in order to verify that the 1D QC is a single-column. Shape-recomputing triggers the whole computation tree, which harms a lot in case of fully-lazy backends (i.e. OmniSci).

I don't think that we want to introduce shape-hint for modin core in the same way as it's done for OmniSci (core shouldn't benefit a lot from this, but maybe you have another opinion?). I'd like to have more general system of handling cache meta-data, something like Index high-level object, which store all the required meta-data cached: names, level-names, length - all of this we can save in cache and provide to the user without actual labels materialization (length is the hardest part, but I think we can retrieve this from the partition's shape cache).

I think we can start from rewriting of how we're computing the frame's shape at the front-end. I believe that getting the frame's length is common operation, but in the current implementation it can't be performed lazily, because it require us to materialize the index:
https://github.com/modin-project/modin/blob/e74cde994bef969abd307ad146e6a5d00847ce51/modin/pandas/base.py#L2896-L2897
We can forward "length-call" to the backend, which will compute the frame shape in the most efficient way (for pandas I believe it would be by summing the partition's shape cache). We can try this as an experiment and look at the perf. changes. @devin-petersohn I can try this if this make sense.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [inline modin/experimental/backends/omnisci/query_compiler.py:89]: Should we add the following check?
```python
if shape_hint is None and len(self._modin_frame.index) == 1:
    shape_hint = "row"
```

Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:89]: No, the meaning of the `shape_hint` is to avoid triggering execution to get information about the frame shape. Accessing the frame's index will trigger the whole delayed computation tree.

**Review by YarShev — APPROVED** [review]
LGTM

**Comment by YarShev** [comment]
@devin-petersohn , @dchigarev , one more thought to the discussion above regarding modin-core. We could add some flag like `default_index`/`default_columns` and get index and columns just summing lengths and widths of the frame (if those should be default) but not calling `get_indices` each time.

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Comment by dchigarev** [comment]
@ienkovich do you have any more comments?

**Comment by ienkovich** [comment]
> @ienkovich do you have any more comments?

The patch is OK for me. Thanks!

**Review by ienkovich — APPROVED** [review]

**Comment by dchigarev** [comment]
@YarShev @vnlitvinov we're all good to merge this?

**Comment by vnlitvinov** [comment]
Let's give others another day to respond, then we can merge.

**Pull request merged**
